File: app/src/main/java/com/zinhao/chtholly/MainActivity.java
package com.zinhao.chtholly;

import android.content.Context;
import android.content.Intent;
import android.content.SharedPreferences;
import android.os.Bundle;
import android.provider.Settings;
import android.text.Editable;
import android.text.TextUtils;
import android.text.TextWatcher;
import android.view.Gravity;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ArrayAdapter;
import android.widget.EditText;
import android.widget.ListView;
import android.widget.TextView;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;
import com.zinhao.chtholly.databinding.ActivityMainBinding;
import com.zinhao.chtholly.entity.Message;
import com.zinhao.chtholly.session.OpenAiSession;
import per.goweii.layer.core.anim.AnimStyle;
import per.goweii.layer.core.widget.SwipeLayout;
import per.goweii.layer.dialog.DialogLayer;

import java.util.List;


public class MainActivity extends AppCompatActivity {
    ActivityMainBinding binding;
    private List<Message> messages;
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        binding = ActivityMainBinding.inflate(getLayoutInflater());
        setContentView(binding.getRoot());

        EditText apiKeyEditText = binding.textInputLayout.getEditText();
        if(apiKeyEditText!=null) {
            apiKeyEditText.setText(BotApp.getInstance().apiKey);
            apiKeyEditText.addTextChangedListener(new TextWatcher() {
                @Override
                public void beforeTextChanged(CharSequence s, int start, int count, int after) {

                }

                @Override
                public void onTextChanged(CharSequence s, int start, int before, int count) {

                }

                @Override
                public void afterTextChanged(Editable s) {
                    BotApp.getInstance().setApiKey(s.toString());
                }
            });
        }

        EditText botNameEditText = binding.textInputLayout2.getEditText();
        if(botNameEditText!=null) {
            botNameEditText.setText(BotApp.getInstance().getBotName());
            botNameEditText.addTextChangedListener(new TextWatcher() {
                @Override
                public void beforeTextChanged(CharSequence s, int start, int count, int after) {

                }

                @Override
                public void onTextChanged(CharSequence s, int start, int before, int count) {

                }

                @Override
                public void afterTextChanged(Editable s) {
                    BotApp.getInstance().setBotName(s.toString());
                }
            });
        }

        EditText adminNameEditText = binding.textInputLayout3.getEditText();
        if(adminNameEditText!=null){
            adminNameEditText.setText(BotApp.getInstance().getAdminName());
            adminNameEditText.addTextChangedListener(new TextWatcher() {
                @Override
                public void beforeTextChanged(CharSequence s, int start, int count, int after) {

                }

                @Override
                public void onTextChanged(CharSequence s, int start, int before, int count) {

                }

                @Override
                public void afterTextChanged(Editable s) {
                    BotApp.getInstance().setAdminName(s.toString());
                }
            });
        }

        binding.toggleButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                startActivity(new Intent(Settings.ACTION_ACCESSIBILITY_SETTINGS));
            }
        });

        binding.button2.setEnabled(false);
        BotApp.getInstance().select(new MessageDao.MessageGetAllListener() {
            @Override
            public void onSuccess(List<Message> result) {
                messages = result;
                runOnUiThread(new Runnable() {
                    @Override
                    public void run() {
                        initMessageDialog();
                    }
                });
            }
        });

        binding.button4.setOnClickListener(this::toVoiceSetting);
    }

    private void toVoiceSetting(View view) {
        Intent i = new Intent(view.getContext(), ServerSettingActivity.class);
        startActivity(i);
    }

    private void initMessageDialog(){
        View dialogContent = getLayoutInflater().inflate(R.layout.bottom_dialog,null,false);
        ListView listView = dialogContent.findViewById(R.id.list);
        listView.setAdapter(new MessageAdapter(this,android.R.layout.simple_list_item_2,messages));
        binding.button2.setEnabled(true);
        binding.button2.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                NekoChatService.getInstance().showFloatWindow();
                new DialogLayer(MainActivity.this)
                        .setContentView(dialogContent)
                        .setGravity(Gravity.BOTTOM)
                        .setSwipeDismiss(SwipeLayout.Direction.BOTTOM)
                        .setBackgroundDimDefault()
                        .setContentAnimator(AnimStyle.BOTTOM)
                        .show();
            }
        });
        binding.button.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                Intent intent = new Intent(v.getContext(),CharacterActivity.class);
                startActivity(intent);
            }
        });
    }

    @Override
    protected void onStop() {
        super.onStop();
        SharedPreferences.Editor editor = BotApp.getInstance().getSharedPreferences().edit();
        editor.putString(BotApp.CONFIG_BOT_NAME,BotApp.getInstance().getBotName());
        editor.putString(BotApp.CONFIG_ADMIN_NAME,BotApp.getInstance().getAdminName());
        editor.putString(BotApp.CONFIG_API_KEY,BotApp.getInstance().apiKey);
        editor.apply();
    }

    @Override
    protected void onResume() {
        super.onResume();
        binding.toggleButton.setChecked(isAccessibilitySettingsOn(this));
        binding.textView.setText(OpenAiSession.getInstance().getChara());
    }

    private boolean isAccessibilitySettingsOn(Context mContext){
        int accessibilityEnabled = 0;
        String service = mContext.getPackageName() + "/" + NekoChatService.class.getCanonicalName();
        try {
            accessibilityEnabled = Settings.Secure.getInt(
                    mContext.getApplicationContext().getContentResolver(),
                    Settings.Secure.ACCESSIBILITY_ENABLED
            );
        } catch (Settings.SettingNotFoundException e) {
            e.printStackTrace();
        }
        TextUtils.SimpleStringSplitter mStringColonSplitter = new TextUtils.SimpleStringSplitter(':');
        if (accessibilityEnabled == 1) {
            String settingValue = Settings.Secure.getString(
                    mContext.getApplicationContext().getContentResolver(),
                    Settings.Secure.ENABLED_ACCESSIBILITY_SERVICES
            );
            mStringColonSplitter.setString(settingValue);
            while (mStringColonSplitter.hasNext()) {
                String accessibilityService = mStringColonSplitter.next();
                if (accessibilityService.equalsIgnoreCase(service)) {
                    return true;
                }
            }
        }
        return false;
    }

    public static class MessageAdapter extends ArrayAdapter<Message>{

        public MessageAdapter(@NonNull Context context, int resource, @NonNull List<Message> objects) {
            super(context, resource, objects);
        }

        @NonNull
        @Override
        public View getView(int position, @Nullable View convertView, @NonNull ViewGroup parent) {
            if(convertView == null)
                convertView = LayoutInflater.from(parent.getContext()).inflate(android.R.layout.simple_list_item_2, parent, false);

            if(convertView!=null){
                TextView tvTitle = convertView.findViewById(android.R.id.text1);
                tvTitle.setText(getItem(position).getSpeaker());

                TextView tvResult = convertView.findViewById(android.R.id.text2);
                tvResult.setText(getItem(position).getMessage());
            }
            assert convertView != null;
            return convertView;
        }
    }
}

File: app/src/main/java/com/zinhao/chtholly/view/MainActivity.kt
package com.zinhao.chtholly.view

import android.content.Intent
import android.os.Bundle
import android.provider.Settings
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.lifecycle.ViewModelProvider
import com.zinhao.chtholly.BotApp
import com.zinhao.chtholly.MainViewModel
import com.zinhao.chtholly.databinding.ActivityMainBinding

class MainActivity : AppCompatActivity() {

    private lateinit var binding: ActivityMainBinding
    private lateinit var viewModel: MainViewModel

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        viewModel = ViewModelProvider(this)[MainViewModel::class.java]
        if(BotApp.getInstance().isFirstRun){
            viewModel.doFirstRun(this)
            startActivity(Intent(this, SetupActivity::class.java))
            finish()
            return
        }
        binding = ActivityMainBinding.inflate(layoutInflater)
        setContentView(binding.root)
        viewModel.checkAccessibilityStatus(this)

        setupObservers()
        setupUI()

        viewModel.refreshCurrentChara()
        viewModel.updateBotName(BotApp.getInstance().botName)
        viewModel.updateAdminName(BotApp.getInstance().adminName)
    }

    private fun setupObservers() {
        // 观察 API Key
        viewModel.apiKey.observe(this) { key ->

        }

        // 观察 Bot 名称
        viewModel.botName.observe(this) { name ->
            binding.tvBotName.setText(name ?: "")
        }

        // 观察管理员名称
        viewModel.adminName.observe(this) { name ->
            setTitle("当前管理员: $name")
        }

        // 观察无障碍服务状态
        viewModel.accessibilityEnabled.observe(this) { enabled ->
            binding.toggleButton.isChecked = enabled
        }

        // 观察当前角色
        viewModel.currentSoul.observe(this) { chara ->
            binding.tvSoul.text = chara
        }

        // 观察 Toast 消息
        viewModel.toastMessage.observe(this) { message ->
            message?.let {
                Toast.makeText(this, it, Toast.LENGTH_SHORT).show()
                viewModel.consumeToastMessage()
            }
        }
    }

    override fun onResume() {
        super.onResume()
        viewModel.refreshCurrentChara()
        viewModel.checkAccessibilityStatus(this)
    }

    private fun setupUI() {
        // 按钮点击事件
        binding.toggleButton.setOnClickListener {
            startActivity(Intent(Settings.ACTION_ACCESSIBILITY_SETTINGS))
        }

        binding.button2.setOnClickListener {
            startActivity(Intent(this, ChatActivity::class.java))
        }

        binding.button5.setOnClickListener {
            viewModel.showFloatWindow(this)
        }

        binding.button7.setOnClickListener {
            startActivity(Intent(this, SetupActivity::class.java))
            finish()
        }

        binding.btSoulEdit.setOnClickListener {
            startActivity(Intent(this, CharacterActivity::class.java))
        }
    }
}